### PR TITLE
Batch reload api to specify what segments to be reloaded on what servers to be more flexible

### PR DESCRIPTION
**Pull request by klsince**
This PR added a new segment reload API to allow more flexible control on segment reloading. The API takes a reloadMap from servers to segments, to specify what segments to be reloaded on what servers. In comparison, the current two APIs either reload a single segment or reload all segments and are not very flexible for scripting. 

e.g. with this new API, one can write a script to reload segments replica group by replica group and do that in small batches to control the reloading workload on servers.

Also extended reload job tracking logic for reloading selected segments on instances.

### CODE REVIEW

**Comment by klsince** [comment]
e.g. when call batchReload with 
```
{"Server_100.107.42.37_7050": ["airlineStats_OFFLINE_16071_16071_0"], "Server_100.107.42.37_7051": ["airlineStats_OFFLINE_16101_16101_1"]}
```

the response is like, one jobId per instance, as each instance got one specific SegmentReloadMsg with a unique msg ID (used as job ID too).
```
{
  "status": "{\"airlineStats_OFFLINE\":{\"Server_100.107.42.37_7050\":{\"reloadJobId\":\"eb021f36-9dc1-4930-8753-426c342c50c9\",\"reloadJobMetaZKStorageStatus\":\"SUCCESS\",\"numMessagesSent\":\"1\"},\"Server_100.107.42.37_7051\":{\"reloadJobId\":\"ddd69f35-cd95-4e94-b34b-de176e2ea3b7\",\"reloadJobMetaZKStorageStatus\":\"SUCCESS\",\"numMessagesSent\":\"1\"}}}"
}
```

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/ControllerJobStatusResource.java [hunk 1]
```diff
@@ -51,45 +54,32 @@ public class ControllerJobStatusResource {
   @ApiOperation(value = "Task status", notes = "Return the status of a given reload job")
   public String reloadJobStatus(@PathParam("tableNameWithType") String tableNameWithType,
       @QueryParam("reloadJobTimestamp") long reloadJobSubmissionTimestamp,
-      @QueryParam("segmentName") String segmentName,
-      @Context HttpHeaders headers)
+      @QueryParam("segmentNames") String segmentNames, @Context HttpHeaders headers)
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/ControllerJobStatusResource.java:57]: This is backward incompatible. Let's just keep the param as is

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java [hunk 2]
```diff
@@ -678,6 +699,81 @@ public SuccessResponse reloadAllSegments(
...
   }
 
+  @POST
+  @Path("segments/{tableName}/batchReload")
+  @Authorize(targetType = TargetType.TABLE, paramName = "tableName", action = Actions.Table.RELOAD_SEGMENT)
+  @Authenticate(AccessType.UPDATE)
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Reload with a map from instances to segments for more flexible control", notes = "Reload "
+      + "with a map from instances to segments for more flexible control")
+  public SuccessResponse reloadSegments(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:709]: Consider integrating this into the table reload API (`POST segments/{tableName}/reload`)

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java [hunk 3]
```diff
@@ -678,6 +699,81 @@ public SuccessResponse reloadAllSegments(
...
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr,
+      @ApiParam(value = "Whether to force server to download segment") @QueryParam("forceDownload")
+      @DefaultValue("false") boolean forceDownload,
+      @ApiParam(value = "Map from instances to segments to reload", required = true) @QueryParam("reloadMap")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:714]: Suggest renaming it to `instanceToSegmentsMap` to be more specific

Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:714]: Consider also allowing reloading a list of segments by adding a `segments` query parameter

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/ControllerJobStatusResource.java (hunk 1 above)
Comment (klsince) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/ControllerJobStatusResource.java:57]: This is a server side API and not used often so I updated it but I’m ok to keep it intact.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java (hunk 3 above)
Comment (klsince) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:714]: The segments should be provided by that Map as well, ie which segments to reload on which instances.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java (hunk 2 above)
Comment (klsince) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:709]: Should be doable. I didn’t try harder as the new batch reload API submits multiple reload jobs to track reload progress on each instance, as each instance may work in different set of segments.

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**